File: battleasia.gg/src/sections/user/play/match-capacity-utils.ts
/** Capacity helpers for admin-set totalPlayer vs live participantsCount. */

export type MatchCapacityInput = {
  participantsCount?: number;
  totalPlayer?: number;
};

export function getMatchCapacityState(match: MatchCapacityInput) {
  const joined = Math.max(0, match.participantsCount ?? 0);
  const max = Math.max(1, match.totalPlayer ?? 100);
  const percent = Math.min(100, (joined / max) * 100);
  const isFull = joined >= max;
  const spotsLeft = Math.max(0, max - joined);

  return { joined, max, percent, isFull, spotsLeft };
}

export function isMatchJoinableByCapacity(match: MatchCapacityInput) {
  return !getMatchCapacityState(match).isFull;
}

File: battleasia-app/lib/core/utils/match_capacity_utils.dart
class MatchCapacityState {
  final int joined;
  final int max;
  final double percent;
  final bool isFull;
  final int spotsLeft;

  const MatchCapacityState({
    required this.joined,
    required this.max,
    required this.percent,
    required this.isFull,
    required this.spotsLeft,
  });
}

MatchCapacityState getMatchCapacityState({
  int? participantsCount,
  int? totalPlayer,
}) {
  final joined = (participantsCount ?? 0).clamp(0, 999999);
  final max = (totalPlayer ?? 100).clamp(1, 999999);
  final percent = (joined / max * 100).clamp(0, 100).toDouble();
  final isFull = joined >= max;
  final spotsLeft = (max - joined).clamp(0, max);

  return MatchCapacityState(
    joined: joined,
    max: max,
    percent: percent,
    isFull: isFull,
    spotsLeft: spotsLeft,
  );
}

bool isMatchJoinableByCapacity({
  int? participantsCount,
  int? totalPlayer,
}) {
  return !getMatchCapacityState(
    participantsCount: participantsCount,
    totalPlayer: totalPlayer,
  ).isFull;
}
